src/FieldValidator.php

```diff
@@ -6,53 +6,47 @@
 
 use ReflectionProperty;
 
-class FieldValidator
+abstract class FieldValidator
 {
-    private static array $typeMapping = [
-        'int' => 'integer',
-        'bool' => 'boolean',
-        'float' => 'double',
-    ];
+    public bool $isNullable;
 
-    private bool $hasTypeDeclaration = false;
+    public bool $isMixed;
 
-    public bool $isNullable = false;
+    public bool $isMixedArray;
 
-    public bool $isMixed = false;
+    public bool $hasDefaultValue;
 
-    public bool $isMixedArray = false;
+    public array $allowedTypes;
 
-    public bool $hasDefaultValue = false;
+    public array $allowedArrayTypes;
 
-    public array $allowedTypes = [];
+    protected static array $typeMapping = [
+        'int' => 'integer',
+        'bool' => 'boolean',
+        'float' => 'double',
+    ];
 
-    public array $allowedArrayTypes = [];
+    protected bool $hasTypeDeclaration;
 
     public static function fromReflection(ReflectionProperty $property): FieldValidator
     {
-        return new self(
-            $property->getDocComment() ?: null,
-            $property->isDefault()
-        );
-    }
+        $docDefinition = null;
 
-    public function __construct(?string $docComment = null, bool $hasDefaultValue = false)
-    {
-        preg_match(
-            '/@var ((?:(?:[\w?|\\\\<>])+(?:\[])?)+)/',
-            $docComment ?? '',
-            $matches
-        );
-
-        $definition = $matches[1] ?? '';
-
-        $this->hasTypeDeclaration = $definition !== '';
-        $this->hasDefaultValue = $hasDefaultValue;
-        $this->isNullable = $this->resolveNullable($definition);
-        $this->isMixed = $this->resolveIsMixed($definition);
-        $this->isMixedArray = $this->resolveIsMixedArray($definition);
-        $this->allowedTypes = $this->resolveAllowedTypes($definition);
-        $this->allowedArrayTypes = $this->resolveAllowedArrayTypes($definition);
+        if ($property->getDocComment()) {
```

Comment:
I deliberately chose to prioritise doccomments over types, in order to still support stuff like this:

```php
    /** @var int[] */
    public array $numbers;
```

If PHP ever adds generics, we can switch the priority of field validators.

Reply:
Purrfect 🐱